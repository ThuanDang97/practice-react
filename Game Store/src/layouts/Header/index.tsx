import { Link, useNavigate } from 'react-router-dom'
import { Box, Drawer, IconButton, Typography, useTheme } from '@mui/material'
import { RiShoppingBag2Line } from 'react-icons/ri'
import PersonIcon from '@mui/icons-material/Person'
import { ChangeEvent, useMemo, useState } from 'react'
import ArrowForwardIcon from '@mui/icons-material/ArrowForward'

// Components
import SearchBar from '@components/SearchBar'
import Button from '@components/Button'
import CartItem from '@components/CartItem'

// Types
import { Game } from '@self-types/Game.types'

// Hooks
import { useAuthContext } from '@hooks/useAuthContext'
import { useCartContext } from '@hooks/useCartContext'

const Header = () => {
  const theme = useTheme()
  const navigate = useNavigate()
  const { userSession, logout } = useAuthContext()
  const { listCart, clearAll, deleteItem } = useCartContext()

  const username = useMemo(
    () =>
      userSession &&
      userSession?.user.firstName + ' ' + userSession?.user.lastName,
    [userSession],
  )

  const [valueSearch, setValueSearch] = useState('')
  const [open, setState] = useState(false)

  const cartQuantity = useMemo(() => listCart.length, [listCart.length])

  const totalPrice = useMemo(
    () =>
      listCart.reduce((acc: number, obj: { price: number }) => {
        return acc + obj.price
      }, 0),
    [listCart],
  )

  const toggleDrawer = () => {
    setState((prev) => !prev)
  }

  // Handle Logout
  const handleLogout = () => {
    logout()
  }

  // Handle login
  const handleClickButtonLogin = () => {
    navigate('/login')
  }

  // Handle change search
  const handleChangeSearch = (e: ChangeEvent<HTMLInputElement>) => {
    setValueSearch(e.target.value)
  }

  // Handle search when user press enter
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (valueSearch !== '' && valueSearch !== undefined) {
      if (e.key === 'Enter') {
        navigate(`/games/search=${valueSearch}`)
      }
    } else {
      navigate('/')
    }
  }

  // Handle search when user press submit button
  const handleSubmitSearch = () => {
    if (valueSearch !== '' && valueSearch !== undefined) {
      navigate(`/games/search=${valueSearch}`)
    } else {
      navigate('/')
    }
  }

  // Clear Cart
  const handleClearCart = async () => {
    await clearAll(userSession?.user.id as string)
  }

  // Handler Deleted Item of Cart
  const handleDeletedItemCart = async (game: Game) => {
    const item = listCart.filter((gameItem) => game !== gameItem)

    await deleteItem(item, userSession?.user.id as string)
  }

  // Handle checkout
  const handleCheckout = () => {
    alert(
      `You will checkout with a product quantity of ${cartQuantity} and a total price of $${totalPrice.toFixed(
        2,
      )}`,
    )
  }

  return (
    <Box
      component="header"
      display="flex"
      alignItems="center"
      justifyContent="space-between"
      height={80}
      padding="0 50px"
      bgcolor={theme.palette.grey[900]}
    >
      <Link to="/">
        <Typography
          component="h1"
          sx={{ color: theme.palette.common.white }}
          fontSize={25}
          fontWeight={700}
          alignItems="center"
          display="flex"
        >
          <Box
            component="img"
            alt="logo"
            src="/images/logo.svg"
            width={20}
            height={20}
            marginRight={1}
          />
          Game Store
        </Typography>
      </Link>
      <SearchBar
        placeholder="Search games..."
        onKeyDown={handleKeyDown}
        onClick={handleSubmitSearch}
        onchange={handleChangeSearch}
      />
      <Box display="flex" alignItems="center" gap={2}>
        {username ? (
          <>
            <Button onClick={handleLogout} variant="contained">
              Logout
            </Button>
            <Typography
              display="flex"
              alignItems="center"
              color={theme.palette.common.white}
            >
              <PersonIcon fontSize="small" />
              {username}
            </Typography>
          </>
        ) : (
          <Button
            onClick={handleClickButtonLogin}
            color="success"
            variant="contained"
          >
            Login
          </Button>
        )}

        <IconButton
          onClick={toggleDrawer}
          size="small"
          sx={{
            color: theme.palette.common.white,
            display: 'flex',
            gap: 1,
            fontWeight: 700,
          }}
        >
          <RiShoppingBag2Line color={theme.palette.common.white} />
          <Typography color={theme.palette.common.white}>Cart</Typography>
          <Box
            bgcolor={theme.palette.common.white}
            borderRadius={12}
            width={20}
            color={theme.palette.common.black}
          >
            <Typography color={theme.palette.common.black}>
              {cartQuantity}
            </Typography>
          </Box>
        </IconButton>
      </Box>

      <Drawer anchor="right" open={open} onClose={toggleDrawer}>
        <Box
          sx={{
            p: 2,
            height: 1,
          }}
          bgcolor={theme.palette.grey[900]}
          width={450}
          position="relative"
        >
          <Box
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            marginTop={2}
            padding="0 12px"
          >
            <Typography
              color={theme.palette.common.white}
              fontWeight={700}
              fontSize={30}
              variant="h3"
            >
              {Number(cartQuantity) !== 0
                ? `${cartQuantity} games`
                : 'No games added'}
            </Typography>
            {Number(cartQuantity) !== 0 && (
              <Button
                onClick={handleClearCart}
                variant="text"
                sx={{
                  fontSize: 18,
                  textTransform: 'capitalize',
                }}
              >
                Clear
              </Button>
            )}
          </Box>
          <Box
            display="flex"
            flexDirection="column"
            marginTop={2}
            alignItems="center"
            gap={1}
          >
            {listCart &&
              listCart.map((game, index) => (
                <CartItem
                  key={game.name + index.toString()}
                  game={game}
                  handleDelete={() => handleDeletedItemCart(game)}
                />
              ))}
          </Box>
          <Box
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            bottom={10}
            position="absolute"
            width="95%"
          >
            <Typography
              color={theme.palette.grey[400]}
              fontWeight={600}
              fontSize={15}
            >
              Total: ${totalPrice.toFixed(2)}
            </Typography>
            <Button
              onClick={handleCheckout}
              variant="text"
              hovercolor={theme.palette.secondary.dark}
              sx={{
                color: theme.palette.common.white,
                fontSize: 24,
                textTransform: 'capitalize',
                fontWeight: 700,
              }}
            >
              Checkout
              <ArrowForwardIcon />
            </Button>
          </Box>
        </Box>
      </Drawer>
    </Box>
  )
}

export default Header
