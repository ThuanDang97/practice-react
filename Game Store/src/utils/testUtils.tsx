import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import { BrowserRouter as Router } from 'react-router-dom'

// Themes
import theme from '@self-types/Themes'

// Contexts
import { AuthProvider } from '@contexts/AuthProvider'
import { CartProvider } from '@contexts/CartProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <AuthProvider>
        <CartProvider>
          <Router>{component}</Router>
        </CartProvider>
      </AuthProvider>
    </ThemeProvider>,
  )
}

export * from '@testing-library/react'
export { customRender as render }
