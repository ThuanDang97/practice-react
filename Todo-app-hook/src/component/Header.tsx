import { useState } from 'react';

type TProps = {
  addTodo: (todo: string) => void;
};

const Header = (props: TProps): JSX.Element => {
  const [text, setText] = useState('');

  const handleGetValue = (e: React.ChangeEvent<HTMLInputElement>) => {
    setText(e.target.value);
  };

  const handleAddTodo = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && text) {
      props.addTodo(text);
      setText('');
    }
  };

  return (
    <header className="header">
      <h1>todos</h1>
      <input
        type="text"
        className="new-todo"
        placeholder="What needs to be done?"
        value={text}
        onChange={handleGetValue}
        onKeyPress={handleAddTodo}
      />
    </header>
  );
};
export default Header;
