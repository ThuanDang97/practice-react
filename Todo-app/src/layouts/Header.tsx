import React, { Component } from 'react';

type Props = {
  addTodo: Function;
};
class Header extends Component<Props> {
  state: {
    text: string;
  };
  constructor(props: Props) {
    super(props);
    this.state = {
      text: '',
    };
  }

  onAddTodo = (e: { key: string }) => {
    if (e.key === 'Enter' && this.state.text) {
      this.props.addTodo(this.state.text);
      this.setState({ text: (this.state.text = '') });
    }
  };

  render(): JSX.Element {
    return (
      <header className="header">
        <h1>todos</h1>
        <input
          type="text"
          className="new-todo"
          placeholder="What needs to be done?"
          value={this.state.text}
          onChange={(e) => this.setState({ text: e.target.value })}
          onKeyPress={(e) => this.onAddTodo(e)}
        />
      </header>
    );
  }
}

export default Header;
