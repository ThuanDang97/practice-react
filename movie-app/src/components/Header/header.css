header {
  display: flex;
  justify-content: space-around;
  align-items: center;
  top: 0;
  left: 0;
  background-image: radial-gradient(at 30% top, #031d33 0%, rgba(3, 37, 65, 1) 70%);
  height: 64px;
  width: 100%;
  z-index: 10;
  transition: top 0.2s linear;
}

.logo {
  margin-right: 16px;
  width: 154px;
  height: 20px;
}

nav ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}
