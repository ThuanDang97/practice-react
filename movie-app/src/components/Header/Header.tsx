import { Link } from 'react-router-dom';
import './header.css';
import logo from '../../assets/logo-blue-short.svg';
import { navHeader } from '../../constants/MockData';
import { Navbar } from '../Navbar/Navbar';

const Header = (): JSX.Element => {
  return (
    <header>
      <Link to="/">
        <h2 className="logo">
          <img src={logo} alt="" />
        </h2>
      </Link>
      <Navbar navItem={navHeader} />
    </header>
  );
};

export { Header };
