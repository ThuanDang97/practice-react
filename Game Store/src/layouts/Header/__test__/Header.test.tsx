// Utils
import { fireEvent, render, screen } from '@utils/testUtils'
import fetchMock from 'jest-fetch-mock'

// Layouts
import Header from '@layouts/Header'

beforeEach(async function () {
  fetchMock.resetMocks()
})

const mockHandleLogout = jest.fn()
const mockHandleLogin = jest.fn()
const mockHandleChangeSearch = jest.fn()
const mockHandleKeyDown = jest.fn()
const mockHandleSubmitSearch = jest.fn()

describe('Header layout', () => {
  jest.mock('@layouts/Header', () => ({
    handleLogout: mockHandleLogout(),
    handleLogin: mockHandleLogin(),
    handleChangeSearch: mockHandleChangeSearch(),
    handleKeyDown: mockHandleKeyDown(),
    handleSubmitSearch: mockHandleSubmitSearch(),
  }))
  it('Header default  matchers DOM Snapshot', () => {
    const { container } = render(<Header />)

    expect(container).toMatchSnapshot()
  })

  it('Header with login account  matchers DOM Snapshot', () => {
    const { container } = render(<Header />)

    expect(container).toMatchSnapshot()
  })

  it('Input shows the correct value entered from the keyboard', () => {
    render(<Header />)

    const input = screen.getByRole('textbox')
    fireEvent.change(input, {
      target: { value: 'GTA' },
    })

    expect(input).toHaveValue('GTA')
  })
})
