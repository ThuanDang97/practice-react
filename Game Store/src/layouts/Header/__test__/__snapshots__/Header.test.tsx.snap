// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header layout Header default  matchers DOM Snapshot 1`] = `
<div>
  <header
    class="MuiBox-root css-zfkp9v"
  >
    <a
      href="/"
    >
      <h1
        class="MuiTypography-root MuiTypography-body1 css-1f869fj-MuiTypography-root"
      >
        <img
          alt="logo"
          class="MuiBox-root css-q0bs1u"
          src="/images/logo.svg"
        />
        Game Store
      </h1>
    </a>
    <div
      class="MuiBox-root css-70qvj9"
    >
      <input
        class="css-1innor3"
        placeholder="Search games..."
        type="text"
        value=""
      />
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17slstj-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiBox-root css-j0ozid"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium css-2navfx-MuiButtonBase-root-MuiButton-root"
        hovercolor="white"
        tabindex="0"
        type="button"
        width="100%"
      >
        Login
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-zbupvj-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          color="#fff"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 22H4a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1h16a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1zm-1-2V4H5v16h14zM9 6v2a3 3 0 0 0 6 0V6h2v2A5 5 0 0 1 7 8V6h2z"
            />
          </g>
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-4peds9-MuiTypography-root"
        >
          Cart
        </p>
        <div
          class="MuiBox-root css-10bwm1q"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-mjamk0-MuiTypography-root"
          >
            0
          </p>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;

exports[`Header layout Header with login account  matchers DOM Snapshot 1`] = `
<div>
  <header
    class="MuiBox-root css-zfkp9v"
  >
    <a
      href="/"
    >
      <h1
        class="MuiTypography-root MuiTypography-body1 css-1f869fj-MuiTypography-root"
      >
        <img
          alt="logo"
          class="MuiBox-root css-q0bs1u"
          src="/images/logo.svg"
        />
        Game Store
      </h1>
    </a>
    <div
      class="MuiBox-root css-70qvj9"
    >
      <input
        class="css-1innor3"
        placeholder="Search games..."
        type="text"
        value=""
      />
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17slstj-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiBox-root css-j0ozid"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium css-2navfx-MuiButtonBase-root-MuiButton-root"
        hovercolor="white"
        tabindex="0"
        type="button"
        width="100%"
      >
        Login
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-zbupvj-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          color="#fff"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 22H4a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1h16a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1zm-1-2V4H5v16h14zM9 6v2a3 3 0 0 0 6 0V6h2v2A5 5 0 0 1 7 8V6h2z"
            />
          </g>
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-4peds9-MuiTypography-root"
        >
          Cart
        </p>
        <div
          class="MuiBox-root css-10bwm1q"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-mjamk0-MuiTypography-root"
          >
            0
          </p>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;
